text/0000-typed-context-injection.md

```diff
@@ -0,0 +1,833 @@
+
+But how does the proposal handle misuse like this:
+
+```rust
+let mut value = 3u32;
+generic_demo::<u32, u32>(Cx::new((&mut value,)));  // Oh no!
+```
+
+Isn't this going to cause undefined behavior in the `generic_demo` function body?
+
+To understand why this is sound, we need to dive deeper into the specific semantics of `Cx`. `Cx` is not an actual type; instead, it's a special type alias resolving to some parametrization of the type `CxRaw<T>`. `CxRaw<T>`, meanwhile, is just a thin newtype wrapper around a tuple of type `T` which is given special treatment by the Rust compiler to allow it to coerce with the semantics described above. `Cx`'s sole role in this scheme is to perform a best-effort deduplication of component types and, like all other type aliases, this deduplication is performed eagerly and is only performed once.
+
+Hence, when a user specifies `Cx<&mut u32, &mut u32>`, this `Cx` alias is transformed into `CxRaw<(&mut u32,)>`. However, if we have function generic with respect to the `T` and `V` type parameters, the parameter `Cx<&mut T, &mut V>` will be resolved to `CxRaw<(&mut T, &mut V)>` and will never change its definition, even if `T` and `V` end up being the same type in a given monomorphization.
+
+```rust
+// This signature:
+fn generics_demo<A, B>(cx: Cx<&mut A, &mut B>) { }
+
+// Is rewritten as:
+fn generics_demo<A, B>(cx: CxRaw<(&mut A, &mut B)>) { }
+
+// Meanwhile, the signature:
+fn non_generic_demo(cx: Cx<&mut u32, &mut u32>) {}
+
+// Is rewritten as:
+fn non_generic_demo(cx: CxRaw<(&mut u32,)>) {}
+
+// This means that the parameter of `generics_demo::<u32, u32>` takes a parameter of the type
+// `CxRaw<(&mut u32, &mut u32)>`, which is distinct from the type of `non_generic_demo`'s parameter,
+// which takes on the type `CxRaw<(&mut u32,)>`.
+```
+
+`CxRaw` does not do any deduplication of its component types. If a component type is duplicated in the source or target `CxRaw` tuple type, the coercion is rejected for being ambiguous since the coercion logic will have no clue where to take references from and where to reborrow them into.
+
+With these two rules in mind, let us now desugar the `generic_demo`:
+
+```rust
+// The user's function...
+fn generic_demo<T, V>(cx: Cx<&mut T, &mut V>) {
+    let a: &mut T = cx;
+    let b: &mut V = cx;
+    let _ = (a, b);
+}
+
+// Desugars to:
+fn generic_demo<T, V>(cx: CxRaw<(&mut T, &mut V)>) {
+    // We can assume that these do not alias because they're stored in different slots in `CxRaw`'s
+	// underlying tuple.
+    let a: &mut T = &mut cx.0.0;
+    let b: &mut V = &mut cx.0.1;
+    let _ = (a, b);
+}
+
+// So when we call it dangerously...
+let mut value = 3u32;
+
+// This coercion is rejected as ambiguous.
+// `CxRaw<(&mut u32,)>` cannot coerce to `CxRaw<(&mut u32, &mut u32)>` because the target
+// contains more than one component of type `&mut u32`.
+generic_demo::<u32, u32>(CxRaw::new((&mut value,)));
```

Comment:
What if I do this:

```rs
fn generic<T: Default, U: Default>() {
    let cx: Cx<T, U> = Cx::new(T::default(), U::default());
    let t: &mut T = cx.extract_mut::<T>();
    let u: &mut U = cx.extract_mut::<U>();
    dbg!(t, u);
}
```

What happens? There is nothing in the function signature that indicates `generic` requires `T != U`. Now we are exposing function implementation to the function signature.

Reply:
Let us desugar that example:

The example:

```rust
fn generic<T: Default, U: Default>() {
    let cx: Cx<T, U> = Cx::new(T::default(), U::default());
    let t: &mut T = cx.extract_mut::<T>();
    let u: &mut U = cx.extract_mut::<U>();
    dbg!(t, u);
}
```

Desugars to:

```rust
fn generic<T: Default, U: Default>() {
    let cx: CxRaw<(T, U)> = CxRaw::new((
        T::default(),
        U::default(),
    ));

    let t: &mut T = &mut *cx.0.0;
    let u: &mut U = &mut *cx.0.1;

    dbg!(t, u);
}
```

...which works, even if `T = V`, so there is no need for bounds anywhere!

Reply:
Then where is the deduplication part? Now `cx` contains two values of the same type.

Reply:
Deduplication only happens once in this entire scheme: in eagerly transforming `Cx` into `CxRaw`. This deduplication exists so we can write:

```rust
// We just have to change this line from a tuple to a `Cx`.
type Descendant1Cx<'a> = Cx<&'a mut SharedSystem, &'a mut Descendant1System>;

fn descendant_1(cx: Descendant1Cx<'_>) {
    ...
}

// Same here!
type Descendant2Cx<'a> = Cx<&'a mut SharedSystem, &'a mut Descendant2System>;

fn descendant_2(cx: Descendant2Cx<'_>) {
    ...
}

// We also need to give the callee its own context.
type CalleeCx<'a> = Cx<Descendant1Cx<'a>, Descendant2Cx<'a>>;

fn callee(cx: CalleeCx<'_>) {
    descendant_1(cx);
    descendant_2(cx);
}

fn caller(
    shared_system: &mut SharedSystem,
    descendant_1_system: &mut Descendant1System,
    descendant_2_system: &mut Descendant2System,
) {
    callee(Cx::new((
        // ...and now we only have to provide one reference to `shared_system`.
        shared_system,
        descendant_1_system,
        descendant_2_system,
    )));
}
```

...without issue, even though `SharedSystem` is, well, shared.

> Now `cx` contains two values of the same type.

That's not a problem since we know exactly where to put these components. `T` goes to `T` and `V` goes to `V` because that's the only valid choice for _all_ situations.

It becomes a problem, however, when the place from which to take or in which to put the component becomes ambiguous. The canonical case is this:

```rust
fn generic_demo<T, V>(cx: Cx<&mut T, &mut V>) {
    let a: &mut T = cx;
    let b: &mut V = cx;
    let _ = (a, b);
}

let mut value = 3u32;
generic_demo::<u32, u32>(Cx::new((&mut value,)));  // Oh no!
```

It's pretty obvious that, when the user tries to extract `T` from their context, they want the component corresponding to `T` and likewise with `V`.

If those components happen to be the same type, however, and you only have a context with one instance of that type, that's where you have a problem. The code:

```rust
let mut value = 3u32;
generic_demo::<u32, u32>(Cx::new((&mut value,)));  // Oh no!
```

...fails because we have no clue where to put that `u32`.

Note that this is different from SFINAE. Substitution errors don't snipe the user from deep inside the substitution graph—they snipe them at the call site as if it were any other generic constraint error.